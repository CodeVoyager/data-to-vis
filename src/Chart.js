import React, {Component} from 'react';
import ReactFauxDOM from 'react-faux-dom'
const d3Selection = require('d3-selection');
const d3Scale = require('d3-scale');
const d3Array = require('d3-array');
const d3TimeFormat = require('d3-time-format');
const d3Axis = require('d3-axis');
const d3Shape = require('d3-shape');
const R = require('ramda');

const DEFAULT_MARGIN = {top: 0, right: 0, bottom: 0, left: 0};
const DEFAULT_WIDTH = 600;
const DEFAULT_HEIGHT = 300;

class Chart extends Component {
    static timeFormat = '%Y-%m-%d';

    constructor () {
        super(...arguments);
        this.state = {
            tooltip: null
        }
    }
    onMouseMove (event) {
        let margin = this.props.margin || DEFAULT_MARGIN;
        let width = (this.props.width || DEFAULT_WIDTH) - margin.left - margin.right;
        let x = d3Scale.scaleTime().rangeRound([0, width]);
        let _data = this.normalizeData(this.props.data);
        let formatTime = d3TimeFormat.timeFormat(Chart.timeFormat);
        let dataElement;
        let selectedDate;
        let leftPosition = event.offsetX - margin.left;
        let tooltip = null;

        x.domain(d3Array.extent(_data, d => d[this.props.xKey]))

        selectedDate = formatTime(x.invert(leftPosition));
        dataElement = R.find(R.propEq('0', selectedDate))(this.props.data);
        if (dataElement) {
            tooltip = {
                x: x.invert(leftPosition),
                y: dataElement[this.props.yKey]
            }
        }
        this.setState(Object.assign({}, this.state, {
            tooltip
        }));
    }

    normalizeData (data) {
        let _data;
        let parseTime = d3TimeFormat.timeParse(Chart.timeFormat);

        if (data[0] instanceof Array) {
            _data = data.map(d => [parseTime(d[this.props.xKey]), d[this.props.yKey]])
        } else if (data[0] instanceof Object) {
            _data = data.map((d) => {
                let newObject = {};
                newObject[this.props.xKey] = parseTime(d[this.props.xKey]);
                newObject[this.props.yKey] = Number(d[this.props.yKey]);

                return Object.assign({}, d, newObject);
            });
        }

        return _data;
    }

    onMouseOver () {
        this.setState(Object.assign({}, this.state, {
            showTooltip: true
        }));
    }

    onMouseOut () {
        this.setState(Object.assign({}, this.state, {
            showTooltip: false
        }));
    }

    render () {
        if (!this.props.data) {
            return null;
        }

        if (R.isNil(this.props.xKey)) {
            throw new Error('xKey is required');
        }

        if (R.isNil(this.props.yKey)) {
            throw new Error('yKey is required');
        }

        let el = ReactFauxDOM.createElement('svg');
        let margin = this.props.margin || DEFAULT_MARGIN;
        let svg = d3Selection.select(el);
        let width = (this.props.width || DEFAULT_WIDTH) - margin.left - margin.right;
        let height = (this.props.height || DEFAULT_HEIGHT) - margin.top - margin.bottom;
        let g = svg.append('g').attr('transform', 'translate(' + margin.left + ',' + margin.top + ')');
        let x = d3Scale.scaleTime().rangeRound([0, width]);
        let y = d3Scale.scaleLinear().rangeRound([height, 0]);
        let line = d3Shape.line()
            .x(d => x(d[this.props.xKey]))
            .y(d => y(d[this.props.yKey]))
        let _data = this.normalizeData(this.props.data);
        let formatTime = d3TimeFormat.timeFormat(Chart.timeFormat);
        let focus;

        svg.attr('width', this.props.width || DEFAULT_WIDTH);
        svg.attr('height', this.props.height || DEFAULT_HEIGHT);

        x.domain(d3Array.extent(_data, d => d[this.props.xKey]))
        y.domain(d3Array.extent(_data, d => d[this.props.yKey]))

        g.append('g')
            .attr('transform', 'translate(0,' + height + ')')
            .call(d3Axis.axisBottom(x));

        g.append('g')
            .call(d3Axis.axisLeft(y));

        g.append('path')
            .datum(_data)
            .attr('fill', 'none')
            .attr('class', 'path')
            .attr('stroke-linejoin', 'round')
            .attr('stroke-linecap', 'round')
            .attr('stroke-width', 1.5)
            .attr('d', line);

        if (this.state.tooltip && this.state.showTooltip) {
            focus = g
                .append('g')
                .attr('class', 'focus')
                .attr("transform", "translate(" + x(this.state.tooltip.x) + "," + y(this.state.tooltip.y) + ")")

            focus.append('circle')
                .attr('r', 4.5);

            focus.append('text')
                .attr('class', 'focus-text')
                .attr('x', 9)
                .attr('dy', '.35em')
                .attr('transform', 'translate(0, -10)')
                .html([this.state.tooltip.y, formatTime(this.state.tooltip.x)].join(' '));
        }

        g.append('rect')
            .attr('class', 'overlay')
            .attr('width', width)
            .attr('height', height)

        let overlay = el.querySelector('.overlay')

        overlay.addEventListener('mouseover', this.onMouseOver.bind(this));
        overlay.addEventListener('mouseout', this.onMouseOut.bind(this));
        overlay.addEventListener('mousemove', this.onMouseMove.bind(this));

        return (
            <div className='graph graph-line'>
                {el.toReact()}
            </div>
        )
    }
}

export default Chart