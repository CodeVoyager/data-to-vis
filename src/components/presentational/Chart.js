import React, {Component} from 'react';
import ReactFauxDOM from 'react-faux-dom'
const d3Selection = require('d3-selection');
const d3Scale = require('d3-scale');
const d3Array = require('d3-array');
const d3TimeFormat = require('d3-time-format');
const d3Axis = require('d3-axis');
const d3Shape = require('d3-shape');
const R = require('ramda');

const DEFAULT_MARGIN = {top: 0, right: 0, bottom: 0, left: 0};
const DEFAULT_WIDTH = 600;
const DEFAULT_HEIGHT = 300;

class Chart extends Component {
    static timeFormat = '%Y-%m-%d';

    constructor () {
        super(...arguments);
        this.state = {
            tooltip: null,
            showTooltip: false
        }
    }

    getMousePosition (event, element) {
        const rect = element.getBoundingClientRect();

        return {
            x: event.clientX - rect.left,
            y: event.clientY - rect.top
        };
    }

    onMouseMove (event) {
        const margin = this.props.margin || DEFAULT_MARGIN;
        const width = (this.props.width || DEFAULT_WIDTH) - margin.left - margin.right;
        const x = d3Scale.scaleTime().rangeRound([0, width]);
        const _data = this.normalizeData(this.props.data);
        const formatTime = d3TimeFormat.timeFormat(Chart.timeFormat);
        const leftPosition = this.getMousePosition(event, event.target).x;
        let selectedDate;
        let tooltip = null;
        let dataElement;

        x.domain(d3Array.extent(_data, d => d[this.props.xKey]))

        selectedDate = formatTime(x.invert(leftPosition));
        dataElement = R.find(R.propEq('0', selectedDate))(this.props.data);
        if (dataElement) {
            tooltip = {
                x: x.invert(leftPosition),
                y: dataElement[this.props.yKey]
            }
        }
        this.setState(Object.assign({}, this.state, {
            tooltip
        }));
    }

    normalizeData (data) {
        let _data;
        const parseTime = d3TimeFormat.timeParse(Chart.timeFormat);

        if (data[0] instanceof Array) {
            _data = data.map(d => [parseTime(d[this.props.xKey]), d[this.props.yKey]])
        } else if (data[0] instanceof Object) {
            _data = data.map((d) => {
                const newObject = {};
                newObject[this.props.xKey] = parseTime(d[this.props.xKey]);
                newObject[this.props.yKey] = Number(d[this.props.yKey]);

                return Object.assign({}, d, newObject);
            });
        }

        return _data;
    }

    onMouseOver () {
        this.setState(Object.assign({}, this.state, {
            showTooltip: true
        }));
    }

    onMouseOut () {
        this.setState(Object.assign({}, this.state, {
            showTooltip: false
        }));
    }

    render () {
        if (!this.props.data) {
            return null;
        }

        if (R.isNil(this.props.xKey)) {
            throw new Error('xKey is required');
        }

        if (R.isNil(this.props.yKey)) {
            throw new Error('yKey is required');
        }

        const el = ReactFauxDOM.createElement('svg');
        const margin = this.props.margin || DEFAULT_MARGIN;
        const svg = d3Selection.select(el);
        const width = (this.props.width || DEFAULT_WIDTH) - margin.left - margin.right;
        const height = (this.props.height || DEFAULT_HEIGHT) - margin.top - margin.bottom;
        const g = svg.append('g').attr('transform', 'translate(' + margin.left + ',' + margin.top + ')');
        const x = d3Scale.scaleTime().rangeRound([0, width]);
        const y = d3Scale.scaleLinear().rangeRound([height, 0]);
        const line = d3Shape.line()
            .x(d => x(d[this.props.xKey]))
            .y(d => y(d[this.props.yKey]))
        const _data = this.normalizeData(this.props.data);
        const formatTime = d3TimeFormat.timeFormat(Chart.timeFormat);
        let focus;

        svg.attr('width', this.props.width || DEFAULT_WIDTH);
        svg.attr('height', this.props.height || DEFAULT_HEIGHT);

        x.domain(d3Array.extent(_data, d => d[this.props.xKey]))
        y.domain(d3Array.extent(_data, d => d[this.props.yKey]))

        g.append('g')
            .attr('transform', 'translate(0,' + height + ')')
            .call(d3Axis.axisBottom(x));

        g.append('g')
            .call(d3Axis.axisLeft(y));

        g.append('path')
            .datum(_data)
            .attr('fill', 'none')
            .attr('class', 'path')
            .attr('stroke-linejoin', 'round')
            .attr('stroke-linecap', 'round')
            .attr('stroke-width', 1.5)
            .attr('d', line);

        if (this.state.tooltip && this.state.showTooltip) {
            focus = g
                .append('g')
                .attr('class', 'focus')
                .attr("transform", "translate(" + x(this.state.tooltip.x) + "," + y(this.state.tooltip.y) + ")");

            focus.append('circle')
                .attr('r', 4.5);

            focus.append('text')
                .attr('class', 'focus-text')
                .attr('x', 9)
                .attr('dy', '.35em')
                .attr('transform', 'translate(0, -10)')
                .html([this.state.tooltip.y, formatTime(this.state.tooltip.x)].join(' '));
        }

        g.append('rect')
            .attr('class', 'overlay')
            .attr('width', width)
            .attr('height', height);

        const overlay = el.querySelector('.overlay');

        overlay.addEventListener('mouseover', this.onMouseOver.bind(this));
        overlay.addEventListener('mouseout', this.onMouseOut.bind(this));
        overlay.addEventListener('mousemove', this.onMouseMove.bind(this));

        return (
            <div className='graph graph-line'>
                {el.toReact()}
            </div>
        )
    }
}

export default Chart