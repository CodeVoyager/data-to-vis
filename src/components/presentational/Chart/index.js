import LineChart from './LineChart';
import React from 'react';

const Chart = (props) => {
    switch (props.type) {
    case 'line':
        return <LineChart {...props}/>;
    default:
        return <LineChart {...props}/>;
    }
}

export {
    LineChart,
}

export default Chart;