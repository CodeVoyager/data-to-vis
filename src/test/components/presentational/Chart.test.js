import React from 'react';
import ReactDOM from 'react-dom';
import Chart from '../../../components/presentational/Chart';
import renderer from 'react-test-renderer';

describe('Chart component', function () {
    const data = [
        ["2013-09-01", 128.2597],
        ["2013-09-02", 127.3648]
    ];
    const xKey = 0;
    const yKey = 1;

    it('renders without crashing', function () {
        const div = document.createElement('div');
        ReactDOM.render(<Chart/>, div);
    });

    it('renders correctly', function () {
        const tree1 = renderer.create(<Chart
            data={data}
            xKey={xKey}
            yKey={yKey}
        />).toJSON();
        expect(tree1).toMatchSnapshot();

        const tree2 = renderer.create(<Chart/>).toJSON();
        expect(tree2).toMatchSnapshot();
    });

    it('throws errors on missing keys', function () {
        let xKey = 'XKEY';
        let yKey = 'YKEY';

        expect(() => {
            renderer.create(<Chart data={data}/>)
        }).toThrowErrorMatchingSnapshot();
        expect(() => {
            renderer.create(<Chart data={data} xKey={null} yKey={yKey}/>)
        }).toThrowErrorMatchingSnapshot();
        expect(() => {
            renderer.create(<Chart data={data} xKey={undefined} yKey={yKey}/>)
        }).toThrowErrorMatchingSnapshot();
        expect(() => {
            renderer.create(<Chart data={data} xKey={xKey} yKey={undefined}/>)
        }).toThrowErrorMatchingSnapshot();
        expect(() => {
            renderer.create(<Chart data={data} xKey={xKey} yKey={null}/>)
        }).toThrowErrorMatchingSnapshot();
    });
});