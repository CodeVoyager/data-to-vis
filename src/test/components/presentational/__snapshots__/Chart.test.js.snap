// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart component renders correctly 1`] = `
<div
  className="graph graph-line"
>
  <svg
    height={300}
    style={Object {}}
    width={600}
  >
    <g
      style={Object {}}
      transform="translate(0,0)"
    >
      <g
        fill="none"
        fontFamily="sans-serif"
        fontSize={10}
        style={Object {}}
        textAnchor="middle"
        transform="translate(0,300)"
      >
        <path
          className="domain"
          d="M0.5,6V0.5H600.5V6"
          stroke="#000"
          style={Object {}}
        />
        <g
          className="tick"
          opacity={1}
          style={Object {}}
          transform="translate(0.5,0)"
        >
          <line
            stroke="#000"
            style={Object {}}
            y2={6}
          />
          <text
            dy="0.71em"
            fill="#000"
            style={Object {}}
            y={9}
          >
            September
          </text>
        </g>
        <g
          className="tick"
          opacity={1}
          style={Object {}}
          transform="translate(75.5,0)"
        >
          <line
            stroke="#000"
            style={Object {}}
            y2={6}
          />
          <text
            dy="0.71em"
            fill="#000"
            style={Object {}}
            y={9}
          >
            03 AM
          </text>
        </g>
        <g
          className="tick"
          opacity={1}
          style={Object {}}
          transform="translate(150.5,0)"
        >
          <line
            stroke="#000"
            style={Object {}}
            y2={6}
          />
          <text
            dy="0.71em"
            fill="#000"
            style={Object {}}
            y={9}
          >
            06 AM
          </text>
        </g>
        <g
          className="tick"
          opacity={1}
          style={Object {}}
          transform="translate(225.5,0)"
        >
          <line
            stroke="#000"
            style={Object {}}
            y2={6}
          />
          <text
            dy="0.71em"
            fill="#000"
            style={Object {}}
            y={9}
          >
            09 AM
          </text>
        </g>
        <g
          className="tick"
          opacity={1}
          style={Object {}}
          transform="translate(300.5,0)"
        >
          <line
            stroke="#000"
            style={Object {}}
            y2={6}
          />
          <text
            dy="0.71em"
            fill="#000"
            style={Object {}}
            y={9}
          >
            12 PM
          </text>
        </g>
        <g
          className="tick"
          opacity={1}
          style={Object {}}
          transform="translate(375.5,0)"
        >
          <line
            stroke="#000"
            style={Object {}}
            y2={6}
          />
          <text
            dy="0.71em"
            fill="#000"
            style={Object {}}
            y={9}
          >
            03 PM
          </text>
        </g>
        <g
          className="tick"
          opacity={1}
          style={Object {}}
          transform="translate(450.5,0)"
        >
          <line
            stroke="#000"
            style={Object {}}
            y2={6}
          />
          <text
            dy="0.71em"
            fill="#000"
            style={Object {}}
            y={9}
          >
            06 PM
          </text>
        </g>
        <g
          className="tick"
          opacity={1}
          style={Object {}}
          transform="translate(525.5,0)"
        >
          <line
            stroke="#000"
            style={Object {}}
            y2={6}
          />
          <text
            dy="0.71em"
            fill="#000"
            style={Object {}}
            y={9}
          >
            09 PM
          </text>
        </g>
        <g
          className="tick"
          opacity={1}
          style={Object {}}
          transform="translate(600.5,0)"
        >
          <line
            stroke="#000"
            style={Object {}}
            y2={6}
          />
          <text
            dy="0.71em"
            fill="#000"
            style={Object {}}
            y={9}
          >
            Mon 02
          </text>
        </g>
      </g>
      <g
        fill="none"
        fontFamily="sans-serif"
        fontSize={10}
        style={Object {}}
        textAnchor="end"
      >
        <path
          className="domain"
          d="M-6,300.5H0.5V0.5H-6"
          stroke="#000"
          style={Object {}}
        />
        <g
          className="tick"
          opacity={1}
          style={Object {}}
          transform="translate(0,288.5)"
        >
          <line
            stroke="#000"
            style={Object {}}
            x2={-6}
          />
          <text
            dy="0.32em"
            fill="#000"
            style={Object {}}
            x={-9}
          >
            127.4
          </text>
        </g>
        <g
          className="tick"
          opacity={1}
          style={Object {}}
          transform="translate(0,255.5)"
        >
          <line
            stroke="#000"
            style={Object {}}
            x2={-6}
          />
          <text
            dy="0.32em"
            fill="#000"
            style={Object {}}
            x={-9}
          >
            127.5
          </text>
        </g>
        <g
          className="tick"
          opacity={1}
          style={Object {}}
          transform="translate(0,221.5)"
        >
          <line
            stroke="#000"
            style={Object {}}
            x2={-6}
          />
          <text
            dy="0.32em"
            fill="#000"
            style={Object {}}
            x={-9}
          >
            127.6
          </text>
        </g>
        <g
          className="tick"
          opacity={1}
          style={Object {}}
          transform="translate(0,188.5)"
        >
          <line
            stroke="#000"
            style={Object {}}
            x2={-6}
          />
          <text
            dy="0.32em"
            fill="#000"
            style={Object {}}
            x={-9}
          >
            127.7
          </text>
        </g>
        <g
          className="tick"
          opacity={1}
          style={Object {}}
          transform="translate(0,154.5)"
        >
          <line
            stroke="#000"
            style={Object {}}
            x2={-6}
          />
          <text
            dy="0.32em"
            fill="#000"
            style={Object {}}
            x={-9}
          >
            127.8
          </text>
        </g>
        <g
          className="tick"
          opacity={1}
          style={Object {}}
          transform="translate(0,121.5)"
        >
          <line
            stroke="#000"
            style={Object {}}
            x2={-6}
          />
          <text
            dy="0.32em"
            fill="#000"
            style={Object {}}
            x={-9}
          >
            127.9
          </text>
        </g>
        <g
          className="tick"
          opacity={1}
          style={Object {}}
          transform="translate(0,87.5)"
        >
          <line
            stroke="#000"
            style={Object {}}
            x2={-6}
          />
          <text
            dy="0.32em"
            fill="#000"
            style={Object {}}
            x={-9}
          >
            128.0
          </text>
        </g>
        <g
          className="tick"
          opacity={1}
          style={Object {}}
          transform="translate(0,54.5)"
        >
          <line
            stroke="#000"
            style={Object {}}
            x2={-6}
          />
          <text
            dy="0.32em"
            fill="#000"
            style={Object {}}
            x={-9}
          >
            128.1
          </text>
        </g>
        <g
          className="tick"
          opacity={1}
          style={Object {}}
          transform="translate(0,20.5)"
        >
          <line
            stroke="#000"
            style={Object {}}
            x2={-6}
          />
          <text
            dy="0.32em"
            fill="#000"
            style={Object {}}
            x={-9}
          >
            128.2
          </text>
        </g>
      </g>
      <path
        className="path"
        d="M0,0L600,300"
        fill="none"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth={1.5}
        style={Object {}}
      />
      <rect
        className="overlay"
        height={300}
        onMouseMove={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={Object {}}
        width={600}
      />
    </g>
  </svg>
</div>
`;

exports[`Chart component renders correctly 2`] = `null`;

exports[`Chart component throws errors on missing keys 1`] = `"xKey is required"`;

exports[`Chart component throws errors on missing keys 2`] = `"xKey is required"`;

exports[`Chart component throws errors on missing keys 3`] = `"xKey is required"`;

exports[`Chart component throws errors on missing keys 4`] = `"yKey is required"`;

exports[`Chart component throws errors on missing keys 5`] = `"yKey is required"`;
